conans/client/rest/download_cache.py

```diff
@@ -0,0 +1,77 @@
+import os
+import shutil
+from threading import Lock
+
+from six.moves.urllib_parse import urlsplit, urlunsplit
+
+from conans.client.tools.files import check_md5, check_sha1, check_sha256
+from conans.errors import ConanException
+from conans.util.files import mkdir
+from conans.util.locks import SimpleLock
+from conans.util.sha import sha256 as sha256_sum
+
+
+class CachedFileDownloader(object):
+
+    def __init__(self, cache_folder, file_downloader, validate_checksum=False):
+        self._cache_folder = cache_folder
+        self._file_downloader = file_downloader
+        self._thread_locks = {}
+        self._validate_checksum = validate_checksum
+
+    def download(self, url, file_path=None, auth=None, retry=None, retry_wait=None, overwrite=False,
+                 headers=None, md5=None, sha1=None, sha256=None):
+        """ compatible interface of FileDownloader + checksum
+        """
+        checksum = sha256 or sha1 or md5
+        h = self._get_hash(url, checksum)
+        lock = os.path.join(self._cache_folder, "locks", h)
+        cached_path = os.path.join(self._cache_folder, h)
+        with SimpleLock(lock):
+            # Once the process has access, make sure multithread is locked too
+            # as SimpleLock doesn't work multithread
+            thread_lock = self._thread_locks.setdefault(lock, Lock())
+            thread_lock.acquire()
+            try:
+                if not os.path.exists(cached_path):
+                    self._file_downloader.download(url, cached_path, auth, retry, retry_wait,
+                                                   overwrite, headers)
+                    if self._validate_checksum:
```

Comment:
I don't understand this comment. The checksum is validated when that file is stored in the cache, and if the file doesn't match the checksum, it is removed from the cache. By definition it is not possible to have an invalid checksum in the cache. Please clarify.

Reply:
Having a closer look at the sources now I realize that the checksum is added to the URL to build the cache key, and only if a checksum is provided the download cache is used, so this should never be an issue.

Nevertheless, I would validate it. I think it is a quick operation (compared to download the file) and it will be the only way to **recover from a corrupted cache**: if the file is not downloaded and the checksum fails, Conan should remove the file and enter the download method again.